pkg/envoy/streamctrl.go

```diff
@@ -38,6 +46,58 @@ type StreamControl struct {
 	sentVersion             uint64
 	currentVersion          uint64
 	currentVersionNackCount uint
+	completions             []versionCompletion
+}
+
+func (ctrl *StreamControl) addCompletion(completions policy.CompletionContainer, msg string) {
+	if completions == nil {
+		return
+	}
+
+	comp, timeout := completions.AddCompletion()
+
+	// Note that we do not start a timer for the timeout, but rely on NACKs to be followed by
+	// retries that allows for checking for timeouts at some times in future. This should
+	// also work accross Envoy restarts.
+	if comp != nil {
+		log.Debug("Envoy: AddCompletion: ", msg)
+		ctrl.completions = append(ctrl.completions,
+			versionCompletion{ctrl.currentVersion, msg, time.Now().Add(timeout), comp})
+	}
+}
+
+// Called with ctrl.cond.L.Lock() held
+func (ctrl *StreamControl) handleCompletions(version uint64, success bool) {
+	now := time.Now()
+
+	for idx := 0; idx < len(ctrl.completions); {
```

Comment:
```
for _, completion := range ctrl.completions {
    ...
}
```

Reply:
Does that support removal of entries in the slice while iterating?